### ISSUE TITLE
Add scope detection based on VCS root

### ISSUE THREAD

**Issue by rdallman** [godoctor/godoctor#41]
i.e. we should be able to crawl up the directory structure until we find a `.git`, `.hg`, etc. (search tools such as "ag" seem to have these kinds of abilities) -- and then set the scope of the refactoring to that. to me it seems a sane default, as long as we can sanely find a vcs root (i.e. if we crawl up to `/` and find one there then it's a little weird / probably wrong). or we could likely use this detection in plugins (without having to write it in each plugin and set it manually). reverting back to how it is now would be easy enough for users to just do `--scope ./`, as opposed to the detection finding the scope dir to use requires a little more machinery. 

I've found myself basically wanting this with regular use of the tool, as in general I'm not starting refactorings in the root directory of a repo, but I want refactorings to propagate across all packages affected in that repo (that may or may not be under my current root). It's of course possible for me to set scope on each refactoring, but it's a tad annoying since in general my editor state of setting the scope is short lived (i.e. I have short vim sessions).

it may also be nice to read in the `.gitignore` and ignore those files, too. I believe we're only loading in any `.go` files laying around so binaries and such are already being ignored, I guess this would handle extra `.go` files laying around that shouldn't be checked into vcs for whatever reason. don't think this is as important as the scope thing.

**Comment by joverbey** [comment]
@rdallman I've been thinking about this for a few weeks.

If there is a single `main` file/function, so we can set the scope to that.  The Go repo, for example, has a `cmd` directory with several different commands, so there are actually several `main` programs in that repo.

If we can find the repo root, it wouldn't be hard to parse through the source files, find all of the files in the `main`, package, and use those to set the scope.  But we'd want to put this on a timeout -- don't search for more than, say, one second.

I haven't touched go/loader in a couple of years and forgot how it works (or doesn't).  Do you have a sense of how well things work if we prime the loader with `main` functions vs. the repo root directory?

I think Rename is the only refactoring where scope really matters, since we look for references across multiple files.  The other refactorings only modify a single file.

**Comment by rdallman** [comment]
This needs to be updated for modules and get rid of the GOPATH lingo, reminding myself to at least update the README instructions. now we can probably crawl up to the root module which is probably the same thing as the git root (just need to get rid of GOPATH to deduce)

**Comment by sorenh** [comment]
Or `go.mod`, whichever comes first?

**Comment by sorenh** [comment]
Even though git most certainly has won the VCS war, it feels wrong to me to have that be a requirement.

**Comment by sorenh** [comment]
Is it really too much to ask for people to run `godoctor` from the root of their workspace, and if they can't, then explicity provide it? That's what gopls does, as far as I can tell.

**Comment by rdallman** [comment]
yea this is more of a convenience, none of this is defaulted or required. we do require setting scope to run a refactoring, I'm mostly referring to updating https://github.com/godoctor/godoctor.vim/#setting-scope-to-git-root in my comment, which is flexible and still allows setting scope per refactoring - I agree yea, want to change lingo to look for go.mod rather than .git once this all works.
